### PR TITLE
adding `NamedArrayPartition` type

### PR DESCRIPTION
**Pull request by jlchan**
I added a `NamedArrayPartition` type in https://github.com/jlchan/StartUpDG.jl; in case it's useful, here are the changes as a PR.

The array type is just an array partition with named fields, e.g., 
```julia
julia> x = NamedArrayPartition(a = randn(10), b = randn(2,2))
julia> x.a # should return the stored result from randn(10)
```
The main difference from `ArrayPartition` is that I require each individual array to have the same `eltype`. I forgot why I did this originally, but I'm happy to fix this if needed.

### CODE REVIEW

**Review by ChrisRackauckas — APPROVED** [review]

File: test/named_array_partition_tests.jl [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    @test typeof(x.a) <: StructVector{<:SVector{2}}
+    @test typeof(x.b) <: StructArray{<:SVector{2}, 2}
+    @test typeof((x->x[1]).(x)) <: NamedArrayPartition
+    @test typeof(map(x->x[1], x)) <: NamedArrayPartition
+end
+
+# x = NamedArrayPartition(a = ones(10), b = rand(20)) 
+# x_ap = ArrayPartition(x)
+# @btime @. x_ap * x_ap; #   498.836 ns (5 allocations: 2.77 KiB) 
+# @btime @. x * x;  # 2.032 μs (5 allocations: 2.84 KiB) - 5x slower than ArrayPartition
```
Comment (jlchan) [inline test/named_array_partition_tests.jl:34]:
```suggestion
```

**Comment by jlchan** [comment]
Sorry, dismissed the review after fixing a conflict

**Comment by ChrisRackauckas** [comment]
bump the version down to test it.

**Comment by jlchan** [comment]
> bump the version down to test it.

Thanks for the tip. Done in https://github.com/SciML/RecursiveArrayTools.jl/pull/293/commits/e6d2680752a0851faafb9fc8d35aa24c3d27b6a5, mind approving CI once more?

**Comment by jlchan** [comment]
Formatting tests were failing for a bunch of files; is it OK if I just lump a bunch of formatting commits into this PR or would you prefer that in a separate one?

**Comment by ChrisRackauckas** [comment]
Don't run the formatter, it needs an update.

File: test/named_array_partition_tests.jl [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    @test typeof(zero(x)) <: NamedArrayPartition
+    @test (y .*= 2).a[1] ≈ 2 # test in-place bcast
+
+    @test length(Array(x))==30
+    @test typeof(Array(x)) <: Array
+    @test propertynames(x) == (:a, :b)
+
+    x = NamedArrayPartition(a = ones(1), b = 2*ones(1))
+    @test Base.summary(x) == string(typeof(x), " with arrays:")
+    @test (@capture_out Base.show(stdout, MIME"text/plain"(), x)) == "(a = [1.0], b = [2.0])"
```
Comment (ChrisRackauckas) [inline test/named_array_partition_tests.jl:19]: What is `@capture_out`?

File: test/named_array_partition_tests.jl [hunk 3]
```diff
@@ -0,0 +1,30 @@
+@testset "NamedArrayPartition tests" begin
+    x = NamedArrayPartition(a = ones(10), b = rand(20))
```
Comment (ChrisRackauckas) [inline test/named_array_partition_tests.jl:2]: missing `using RecursiveArrayTools, Test` on top

File: test/named_array_partition_tests.jl [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    @test typeof(zero(x)) <: NamedArrayPartition
+    @test (y .*= 2).a[1] ≈ 2 # test in-place bcast
+
+    @test length(Array(x))==30
+    @test typeof(Array(x)) <: Array
+    @test propertynames(x) == (:a, :b)
+
+    x = NamedArrayPartition(a = ones(1), b = 2*ones(1))
+    @test Base.summary(x) == string(typeof(x), " with arrays:")
+    @test (@capture_out Base.show(stdout, MIME"text/plain"(), x)) == "(a = [1.0], b = [2.0])"
```
Comment (ChrisRackauckas) [inline test/named_array_partition_tests.jl:21]: Just show to an IOBuffer

File: test/named_array_partition_tests.jl (hunk 2 above)
Comment (jlchan) [reply test/named_array_partition_tests.jl:21]: Resolving, I'll replace `@capture_out` with an IOBuffer as advised below

File: test/named_array_partition_tests.jl (hunk 3 above)
Comment (jlchan) [reply test/named_array_partition_tests.jl:4]: Shoot - thank you for catching that.

File: test/named_array_partition_tests.jl (hunk 4 above)
Comment (jlchan) [reply test/named_array_partition_tests.jl]: Thanks; done in https://github.com/SciML/RecursiveArrayTools.jl/pull/293/commits/18a75d003ce48571b9026141219097b38ce0b107

**Comment by jlchan** [comment]
Thanks for catching that I forgot to load the packages in my tests @ChrisRackauckas. I think I've fixed the last failing test - would you mind approving the CI again?

**Comment by ChrisRackauckas** [comment]
Last thing, just add to the docs.

**Comment by jlchan** [comment]
Thanks!

**Pull request merged**